net/dns/resolver/forwarder.go

```diff
@@ -100,178 +89,149 @@ func getTxID(packet []byte) txid {
 }
 
 func (f *forwarder) setRoutes(routes []route) {
 	f.mu.Lock()
+	defer f.mu.Unlock()
 	f.routes = routes
-	f.mu.Unlock()
 }
 
 // send sends packet to dst. It is best effort.
-func (f *forwarder) send(packet []byte, dst netaddr.IPPort) {
-	connIdx := rand.Intn(connCount)
-	conn := f.conns[connIdx]
-	conn.send(packet, dst)
-}
-
-func (f *forwarder) recv(conn *fwdConn) {
-	defer f.wg.Done()
-
-	for {
-		select {
-		case <-f.closed:
-			return
-		default:
-		}
-		// The 1 extra byte is to detect packet truncation.
-		out := make([]byte, maxResponseBytes+1)
-		n := conn.read(out)
-		var truncated bool
-		if n > maxResponseBytes {
-			n = maxResponseBytes
-			truncated = true
-		}
-		if n == 0 {
-			continue
-		}
-		if n < headerBytes {
-			f.logf("recv: packet too small (%d bytes)", n)
-		}
+//
+// send expects the reply to have the same txid as txidOut.
+//
+// The provided closeOnCtxDone lets send register values to Close if
+// the caller's ctx expires. This avoids send from allocating its own
+// waiting goroutine to interrupt the ReadFrom, as memory is tight on
+// iOS and we want the number of pending DNS lookups to be bursty
+// without too much associated goroutine/memory cost.
+func (f *forwarder) send(ctx context.Context, txidOut txid, closeOnCtxDone *closePool, packet []byte, dst netaddr.IPPort) ([]byte, error) {
+	// TODO(bradfitz): if dst.IP is 8.8.8.8 or 8.8.4.4 or 1.1.1.1, etc, or
+	// something dynamically probed earlier to support DoH or DoT,
+	// do that here instead.
+
+	var linkName string
+	if f.linkSel != nil {
+		linkName = f.linkSel.PickLink(dst.IP())
+	}
 
-		out = out[:n]
-		txid := getTxID(out)
+	// Make a new connection.
+	var lc net.ListenConfig
+	if initListenConfig != nil && linkName != "" {
+		initListenConfig(&lc, f.linkMon, linkName)
+	}
+	conn, err := lc.ListenPacket(ctx, "udp", ":0")
+	if err != nil {
+		f.logf("ListenPacket failed: %v", err)
+		return nil, err
+	}
+	defer conn.Close()
 
-		if truncated {
-			const dnsFlagTruncated = 0x200
-			flags := binary.BigEndian.Uint16(out[2:4])
-			flags |= dnsFlagTruncated
-			binary.BigEndian.PutUint16(out[2:4], flags)
+	closeOnCtxDone.Add(conn)
+	defer closeOnCtxDone.Remove(conn)
 
-			// TODO(#2067): Remove any incomplete records? RFC 1035 section 6.2
-			// states that truncation should head drop so that the authority
-			// section can be preserved if possible. However, the UDP read with
-			// a too-small buffer has already dropped the end, so that's the
-			// best we can do.
+	if _, err := conn.WriteTo(packet, dst.UDPAddr()); err != nil {
+		if err := ctx.Err(); err != nil {
+			return nil, err
 		}
+		return nil, err
+	}
 
-		f.mu.Lock()
-
-		record, found := f.txMap[txid]
-		// At most one nameserver will return a response:
-		// the first one to do so will delete txid from the map.
-		if !found {
-			f.mu.Unlock()
-			continue
+	// The 1 extra byte is to detect packet truncation.
+	out := make([]byte, maxResponseBytes+1)
+	n, _, err := conn.ReadFrom(out)
```

Comment:
Should there be a `conn.SetDeadline()` so that the `WriteTo`/`ReadFrom` will abort at the same time as the context's timeout?

Reply:
The closeOnCtxDone handles that, and also handles other reasons a context might be done (e.g. explicit context.WithCancel CancelFunc being called).

Reply:
Ohh, right, it'll return on something else calling Close. OK, nevermind.